src/main/kotlin/net/maizegenetics/phgv2/api/TileDBVcfReader.kt

```diff
@@ -0,0 +1,344 @@
+                contigs
+            )
+
+            //get startPos (pos_start, INT32)
+            decodeInt(
+                dbReader.getBuffer("pos_start"),
+                numberOfRecords,
+                startPositions
+            )
+
+            //get endPos (pos_end, INT32)
+            decodeInt(
+                dbReader.getBuffer("pos_end"),
+                numberOfRecords,
+                endPositions
+            )
+
+            //get genotype (alleles, varlen list of CHAR; fmt_GT, varlen, nullable, INT32)
+            //get alleles and fmt_GT, then derive genotypes
+            val gtList = mutableListOf<List<Int>>()
+            val alleles = mutableListOf<List<String>>()
+
+            decodeVarlenNullableInt(dbReader.getBuffer("fmt_GT"),
+                dbReader.getOffsets("fmt_GT"),
+                BitSet.valueOf(dbReader.getBitMap("fmt_GT")),
+                numberOfRecords,
+                gtList
+            )
+
+            decodeVarlenListChar(dbReader.getBuffer("alleles"),
+                dbReader.getOffsets("alleles"),
+                dbReader.getListOffsets("alleles"),
+                numberOfRecords,
+                alleles
+            )
+
+            //get AD (fmt_AD, varlen, nullable, INT32)
+            val adList = mutableListOf<List<Int>>()
+            if (hasAD) {
+                decodeVarlenNullableInt(dbReader.getBuffer("fmt_AD"),
+                    dbReader.getOffsets("fmt_AD"),
+                    BitSet.valueOf(dbReader.getBitMap("fmt_AD")),
+                    numberOfRecords, adList
+                )
+            }
+
+            for (ndx in 0 until numberOfRecords) {
+                val currAlleles = alleles[ndx]
+                val geno = gtList[ndx].map { if( it < 0) "." else currAlleles[it] }
+                genotypes.add(geno)
+                if (hasAD) {
+                    val ad = gtList[ndx].map { if( it < 0) -1 else adList[ndx][it] }
+                    ADs.add(ad)
+                }
+
+            }
+
+
+            //get DP (fmt_DP, nullable, INT32)
+            if (hasDP) {
+                decodeNullableInt(dbReader.getBuffer("fmt_DP"),
+                    BitSet.valueOf(dbReader.getBitMap("fmt_DP")),
+                    numberOfRecords,
+                    DPs
+                )
+            }
+
+        }
+
+        //convert values to a list of sample data
+        return sampleNames.indices.map { SampleData(sampleNames[it], contigs[it],
+            startPositions[it], endPositions[it], genotypes[it],
+            if (hasAD) ADs[it] else null, if (hasDP) DPs[it] else null) }
+
+    }
+
+    private fun decodeVarlenChar(dataBuffer: ByteBuffer,
+                         offsetBuffer: ByteBuffer,
+                         numberOfRecords: Int,
+                         outputList: MutableList<String>)  {
+        val offsets = IntArray(numberOfRecords + 1)
+        offsetBuffer.asIntBuffer().get(offsets)
+
+        dataBuffer.rewind()
```

Comment:
Why is rewind() needed here but not elsewhere?  I see we call decodeVarlenChar() with dataBuffer equal to:

  dbReader.getBuffer("sample_name"). or
   dbReader.getBuffer("contig").  or
   dbReader.getBuffer("alleles") (from the decodeVarlenListChar() call)
   
   In each of these cases I didn't see the databuffer being read before the decodeVaLenChar, but perhaps I missed where that happens

Reply:
Rewind sets the data buffer pointer to the start of the buffer. If the buffer is smaller than the data, then after the buffers are processed, query.submit is run again. If there is no buffer.rewind, the pointer stays where it was after the last loop. Since this method does not know if it has been run before, its best to use rewind just in case. All the other buffer reading methods use the asIntBuffer method before retrieving data. That returns an IntBuffer set at the beginning, so rewinding is not needed.

Reply:
Thanks for the explanation.